### PR TITLE
Add support for @classpath_abi macro

### PR DESCRIPTION
**Pull request by raviagarwal7**
This is similar to @classpath macro but gives the corresponding 
class-abi jar for the classpath entries.

This is useful for genrules which only depends on the dependencies abi’s.
For eg: android lint where it need not run the lint checks again if 
the dependent’s abi doesn’t change.

Usage: `$(@classpath_abi :<target_path>)`

### CODE REVIEW

**Comment by raviagarwal7** [comment]
Added more context in the description and also added to the macros soy.
Have fixed the build errors as well.

File: src/com/facebook/buck/jvm/core/HasJavaAbi.java [hunk 1]
```diff
@@ -43,18 +43,18 @@
   Flavor SOURCE_ONLY_ABI_FLAVOR = InternalFlavor.of("source-only-abi");
   Flavor VERIFIED_SOURCE_ABI_FLAVOR = InternalFlavor.of("verified-source-abi");
 
-  static BuildTarget getClassAbiJar(BuildTarget libraryTarget) {
```
Comment (styurin) [inline src/com/facebook/buck/jvm/core/HasJavaAbi.java]: Any changes in this file? Why do you need to move this method?

File: src/com/facebook/buck/rules/macros/ClasspathMacroExpander.java [hunk 2]
```diff
@@ -85,7 +85,7 @@ protected Arg expand(SourcePathResolver resolver, ClasspathMacro ignored, BuildR
   // The # characters that might be present in classpaths due to flavoring would be read as
   // comments. As a simple workaround, we quote the entire classpath.
   private static class ClassPathWriteToFileArg extends WriteToFileArg {
-    public ClassPathWriteToFileArg(BuildTarget target, String prefix, Arg delegate) {
```
Comment (styurin) [inline src/com/facebook/buck/rules/macros/ClasspathMacroExpander.java]: Changes in this file look unrelated, could you revert them?

File: src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    BuildRule inputRule = resolve(resolver, input);
+    return expand(resolver, inputRule);
+  }
+
+  protected Arg expand(BuildRuleResolver ruleResolver, BuildRule inputRule) throws MacroException {
+
+    ImmutableList<SourcePath> jarPaths =
+        getHasClasspathEntries(inputRule)
+            .getTransitiveClasspathDeps()
```
Comment (raviagarwal7) [inline src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java]: `getTransitiveClasspathDeps` returns both `#aar_prebuilt_jar,class-abi` and `#class-abi` flavours. I verified that both the output jar's are same, just on different paths.

//cache:com.support-compat-27.0.2.aar#aar_prebuilt_jar,class-abi
buck-out/gen/cache/com.support-27.0.2.aar#aar_prebuilt_jar,class-abi/com.support-27.0.2.aar-abi.jar

//cache:com.support-compat-27.0.2.aar#class-abi
buck-out/gen/cache/com.support-27.0.2.aar#class-abi/com.support-27.0.2.aar-abi.jar

How can I filter and keep only the `#class-abi` deps? Also is there a scenario where the output of the two flavours will differ?

File: src/com/facebook/buck/jvm/core/HasJavaAbi.java (hunk 1 above)
Comment (raviagarwal7) [reply src/com/facebook/buck/jvm/core/HasJavaAbi.java]: just moved the method to be in accordance with others, but will revert back

File: src/com/facebook/buck/rules/macros/ClasspathMacroExpander.java (hunk 2 above)
Comment (raviagarwal7) [reply src/com/facebook/buck/rules/macros/ClasspathMacroExpander.java]: Intellij was showing suggestions that this is not needed, will revert back.

File: src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java (hunk 3 above)
Comment (raviagarwal7) [reply src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java]: added a filter to only include rules which have non-null `sourcePathToOutput`, same behaviour is in `ClasspathMacro`.

File: src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  /**
+   * Get the class abi jar if present for the rule otherwise return the rule's output
+   *
+   * @param rule The rule whose jar path needs to be returned
+   * @param ruleResolver BuildRuleResolver
+   * @return class abi jar or output jar if not found
+   */
+  private SourcePath getJarPath(BuildRule rule, BuildRuleResolver ruleResolver) {
```
Comment (styurin) [inline src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java]: This needs to be annotated with `@Nullable`.

File: test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            .addSrc(Paths.get("Example.java")) // Force a jar to be created
+            .build(ruleResolver, filesystem);
+
+    assertExpandsTo(
+        rule,
+        ruleResolver,
+        String.format(
+            "%s" + File.separator + "%s",
+            filesystem.getRootPath(),
+            "buck-out/gen/cheese/cake#class-abi/cake-abi.jar"));
```
Comment (styurin) [inline test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java]: These tests are failing on Windows. Our CI doesn't run these tests for some reason (we're investigating), but it looks like it's failing because of the path separators.

File: src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java (hunk 4 above)
Comment (raviagarwal7) [reply src/com/facebook/buck/rules/macros/ClasspathAbiMacroExpander.java]: done

File: test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java (hunk 5 above)
Comment (raviagarwal7) [reply test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java]: using `new File.("file-path-string")toPath()` now, hopefully this would work fine on windows.

File: test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+import org.hamcrest.Matchers;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ClasspathAbiMacroExpanderTest {
+  private ClasspathAbiMacroExpander expander;
+  private ProjectFilesystem filesystem;
+
+  @Before
+  public void createMacroExpander() {
```
Comment (styurin) [inline test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java]: this is usually called `setUp`. It also does more than creating macro expander.

File: test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    this.expander = new ClasspathAbiMacroExpander();
+    this.filesystem = FakeProjectFilesystem.createJavaOnlyFilesystem();
+  }
+
+  @Test
+  public void shouldIncludeARuleIfNothingIsGiven() throws Exception {
```
Comment (styurin) [inline test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java]: test methods usually start with `test` so that the method name is a verb, right now it's a statement

File: test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java (hunk 6 above)
Comment (raviagarwal7) [reply test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java]: done

File: test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java (hunk 7 above)
Comment (raviagarwal7) [reply test/com/facebook/buck/rules/macros/ClasspathAbiMacroExpanderTest.java]: done

**Comment by kageiit** [comment]
cc @styurin the comments are addressed. Any update on this?
